Build limit group

Coronia opened the pull request:
Different technos can now share build limit within a group. Originally implemented by ststl. I've gotten the permission to transplant it to mainline Phobos and make further enhancement on it.

- `BuildLimitGroup.Types` determines the technos that'll be used for build limit conditions of the selected techno.
- `BuildLimitGroup.Nums` determines the amount of technos that would reach the build limit. If using a single integer, it'll use the sum of all technos in the group to calculate build limit. If using a list of integers with the same size of `BuildLimitGroup.Types`, it'll calculate build limit per techno.
- `BuildLimitGroup.ContentIfAnyMatch` determines the rule of calculating build limit per techno. If set to true, build limit will be content if the amount of any techno in the group reaches its `BuildLimitGroup.Nums` value. If set to false, then it'll only be content if the amount of all technos in the group reached.
- `BuildLimitGroup.NotBuildableIfQueueMatch` determines the moment to stop the techno's production. If set to true, its production will be stopped once the condition is content by the sum of real technos and technos in production queue. If set to false, it'll only be stopped when the condition is content by real technos.

In `rulesmd.ini`:
```ini
[SOMETECHNO]                                    ; TechnoType
BuildLimitGroup.Types=                          ; list of TechnoType names
BuildLimitGroup.Nums=                           ; integer, or a list of integers
BuildLimitGroup.ContentIfAnyMatch=false         ; boolean
BuildLimitGroup.NotBuildableIfQueueMatch=false  ; boolean
```

Starkku commented:
I'll do a full code review later but I tested this in both skirmish and MP and it appeared to work fine, no crashes and no desyncs.

FS-21 commented on src/Ext/House/Body.cpp:
```diff
@@ -622,3 +622,75 @@ DEFINE_HOOK(0x50114D, HouseClass_InitFromINI, 0x5)
+				if (ownedNow < pItemExt->BuildLimit_Group_Limits[i]
+				|| includeQueued && FactoryClass::FindByOwnerAndProduct(pThis, pType))
+					return CanBuildResult::Buildable;
+			}
+			return CanBuildResult::TemporarilyUnbuildable;
```
```suggestion
			}

			return CanBuildResult::TemporarilyUnbuildable;
```

FS-21 commented on src/Ext/House/Hooks.BuildLimit.cpp:
```diff
@@ -0,0 +1,156 @@
+{
+	const auto pTypeExt = TechnoTypeExt::ExtMap.Find(pType);
+
+	if (pTypeExt->BuildLimit_Group_Types.empty()
+		|| pTypeExt->BuildLimit_Group_Limits.empty())
+		return false;
```
multiline "if" -> put brackets

FS-21 commented on src/Ext/House/Hooks.BuildLimit.cpp:
```diff
@@ -0,0 +1,156 @@
+			if (!ignoreQueued)
+				queued += QueuedNum(pHouse, pTmpType);
+
+			count += pHouse->CountOwnedNow(pTmpType);
+			if (pTmpType == pType)
```
```suggestion
			count += pHouse->CountOwnedNow(pTmpType);

			if (pTmpType == pType)
```

FS-21 commented on src/Ext/House/Hooks.BuildLimit.cpp:
```diff
@@ -0,0 +1,156 @@
+			{
+				if (pTypeExt->BuildLimit_Group_Any)
+				{
+					if (num >= 0 || pTypeExt->BuildLimit_Group_Stop)
+						RemoveProduction(pHouse, pType, -1);
+					return true;
```
```suggestion
					if (num >= 0 || pTypeExt->BuildLimit_Group_Stop)
						RemoveProduction(pHouse, pType, -1);

					return true;
```

FS-21 commented on src/Ext/House/Hooks.BuildLimit.cpp:
```diff
@@ -0,0 +1,156 @@
+	GET(const HouseClass*, pThis, ECX);
+	GET_STACK(const TechnoTypeClass*, pType, 0x4);
+	GET(CanBuildResult, aresResult, EAX);
+
+
```
remove 1 empty line

Starkku commented on src/Ext/TechnoType/Body.cpp:
```diff
@@ -295,6 +295,11 @@ void TechnoTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
 	this->AttachEffect_InitialDelays.Read(exINI, pSection, "AttachEffect.InitialDelays");
 	this->AttachEffect_RecreationDelays.Read(exINI, pSection, "AttachEffect.RecreationDelays");
 
+	this->BuildLimit_Group_Types.Read(exINI, pSection, "BuildLimitGroup.Types");
+	this->BuildLimit_Group_Any.Read(exINI, pSection, "BuildLimitGroup.ContentIfAnyMatch");
+	this->BuildLimit_Group_Limits.Read(exINI, pSection, "BuildLimitGroup.Nums");
+	this->BuildLimit_Group_Stop.Read(exINI, pSection, "BuildLimitGroup.NotBuildableIfQueueMatch");
```
Internal field names and INI key names should match as closely as possible unless there's a good reason to not. I would remove the underscore between Limit and Group as well.

Starkku reviewed (changes requested):
Looks mostly fine, couple of minor code style nitpicks though.

chaserli commented on src/Ext/House/Hooks.BuildLimit.cpp:
```diff
@@ -0,0 +1,153 @@
+
+#include <Ext/TechnoType/Body.h>
+
+inline int __fastcall QueuedNum(const HouseClass* pHouse, const TechnoTypeClass* pType)
```
any reason for _fastcall here?

chaserli commented on src/Ext/House/Hooks.BuildLimit.cpp:
```diff
@@ -0,0 +1,153 @@
+	return false;
+}
+
+DEFINE_HOOK(0x4F8361, HouseClass_CanBuild, 0x5)
```
duplication, you could've done it already

chaserli commented on src/Ext/House/Body.cpp:
```diff
@@ -622,3 +622,73 @@ DEFINE_HOOK(0x50114D, HouseClass_InitFromINI, 0x5)
+int HouseExt::CountOwnedIncludeDeploy(const HouseClass* pThis, const TechnoTypeClass* pItem)
+{
+	int count = pThis->CountOwnedNow(pItem);
+	count += pItem->DeploysInto == nullptr ? 0 : pThis->CountOwnedNow(pItem->DeploysInto);
```
explicitly check if a pointer == nullptr seems not conforming to our codestyle, same thing for all similar references

chaserli commented on src/Ext/House/Body.cpp:
```diff
@@ -622,3 +622,73 @@ DEFINE_HOOK(0x50114D, HouseClass_InitFromINI, 0x5)
+		for (size_t i = 0; i < std::min(pItemExt->BuildLimitGroup_Types.size(), pItemExt->BuildLimitGroup_Nums.size()); i++)
+		{
+			TechnoTypeClass* pType = pItemExt->BuildLimitGroup_Types[i];
+			int ownedNow = CountOwnedIncludeDeploy(pThis, pType);
```
```suggestion
			int ownedNow = HouseExt::CountOwnedIncludeDeploy(pThis, pType);
```
same to other references

chaserli commented on src/Ext/House/Hooks.BuildLimit.cpp:
```diff
@@ -0,0 +1,153 @@
+	return 0;
+}
+
+DEFINE_HOOK(0x50B669, HouseClass_ShouldDisableCameo, 0x5)
```
similar issue, either move that one here or move this one there

Coronia commented:
moved all hooks into Hooks.Update and removed Hooks.BuildLimits, tho I didn't find out a duplicate of hook 0x50B669 yet

I've asked ststl and made sure that __fastcall can be removed here. Also fixed other code style issues

Starkku commented:
Why is non-upgrade related code in `Hooks.Upgrade.cpp` now?

Coronia commented:
According to the suggestion of MigEater I've added these additional functions:

- `BuildLimitGroup.Factor` determines each of this techno instance will count as what value when calculating build limit.
  - This is only used by BuildLimitGroup. No other place will use it when counting owned objects, including `BuildLimitGroup.ExtraLimit`.
- You can also add an extra value into `BuildLimitGroup.Nums`, determined by the amount of specific technos owned by its house.
  - `BuildLimitGroup.ExtraLimit.Types` determines the technos that'll be used for extra value calculation.
  - `BuildLimitGroup.ExtraLimit.Nums` determines the actual value of increment. Value matching the position in `BuildLimitGroup.ExtraLimit.Types` is used for that type. For each of these technos, it'll increase the extra value by its amount * corresponding value from the list.
  - `BuildLimitGroup.ExtraLimit.MaxCount` determines the maximum amount of technos being counted into the extra value calculation. Value matching the position in `BuildLimitGroup.ExtraLimit.Types` is used for that type. If not set or set to a value below 1, it'll be considered as no maximum count.

  In `rulesmd.ini`:
```ini
[SOMETECHNO]                                    ; TechnoType
BuildLimitGroup.Factor=1                        ; integer
BuildLimitGroup.ExtraLimit.Types=               ; list of TechnoType names
BuildLimitGroup.ExtraLimit.Nums=                ; list of integers
BuildLimitGroup.ExtraLimit.MaxCount=            ; list of integers
```

Please test and review it when you have time, cuz I'm not 100% sure my calculation logic is correct

Coronia commented on src/Ext/BuildingType/Hooks.Upgrade.cpp:
```diff
@@ -103,6 +258,14 @@ DEFINE_HOOK(0x4F8361, HouseClass_CanBuild_UpgradesInteraction, 0x5)
+	if (resultOfAres == CanBuildResult::Buildable)
+		R->EAX(HouseExt::BuildLimitGroupCheck(pThis, pItem, buildLimitOnly, includeInProduction));
+	else
+		return 0;
```
it is, the following steps will be executed only if the Ares result is buildable

CrimRecya commented on src/Ext/House/Body.cpp:
```diff
@@ -622,3 +622,259 @@ DEFINE_HOOK(0x50114D, HouseClass_InitFromINI, 0x5)
 	return 0;
 }
+
+int CountOwnedIncludeDeploy(const HouseClass* pThis, const TechnoTypeClass* pItem)
```
```suggestion
int HouseExt::CountOwnedIncludeDeploy(const HouseClass* pThis, const TechnoTypeClass* pItem)
```
Should it be like this?

Coronia replied:
Now that all functions are moved into HouseClass, there's no need to make this a HouseExt function for outer call

chaserli reviewed (approved):
codestyle lgtm, impl needs testing in action
